tests/l3_test.c

```diff
@@ -0,0 +1,185 @@
+#include "l3_test.h"
+
+#include <errno.h>
+#include <fcntl.h>
+#include <sys/mman.h>
+#include <unistd.h>
+
+/**
+ * Support for different 'logging' types under L3's APIs.
+ */
+static const char * L3_logtype_name[] = {
+                          "L3_LOG_unknown"
+                        , "L3_LOG_MMAP"
+                        , "L3_LOG_FPRINTF"
+                        , "L3_LOG_WRITE"
+                        , "L3_LOG_WRITE_MSG"
+                };
+
+/**
+ * l3_test_logtype_name() - Map log-type ID to its name.
+ */
+static const char *
+l3_test_logtype_name(l3_test_log_t logtype)
+{
+    return (((logtype < 0) || (logtype >= L3_LOGTYPE_MAX))
+             ? L3_logtype_name[L3_LOG_UNDEF]
+             : L3_logtype_name[logtype]);
+}
+
+
+L3_STATIC_ASSERT((L3_ARRAY_LEN(L3_logtype_name) == L3_LOGTYPE_MAX),
+                  "Incorrect size of array L3_logtype_name[]");
+
+static FILE *  l3_test_log_fh = NULL;   // L3_LOG_FPRINTF: Opened by fopen()
+
+static int     l3_log_fd = -1;     // L3_LOG_WRITE: Opened by open()
+
+/**
+ * ****************************************************************************
+ * Initialize L3's logging sub-system to use write() to named `path`.
+ */
+static int
+l3_init_write(const char *path)
+{
+    if (!path) {
+        fprintf(stderr, "%s: Invalid arg 'path'=%p\n", __func__, path);
+        return -1;
+    }
+
+    l3_log_fd = open(path, (O_RDWR | O_CREAT | O_TRUNC), 0666);
+    if (l3_log_fd == -1) {
+        fprintf(stderr, "%s: Error opening log-file at '%p'. Error=%d\n",
+                __func__, path, errno);
+        return -1;
+    }
+    printf("Initialized write() logging to '%s'\n", path);
+    return 0;
+}
+
+/**
+ * ****************************************************************************
+ * Initialize L3's logging sub-system to use fprintf() to named `path`.
+ */
+static int
+l3_init_fprintf(const char *path)
+{
+    if (!path) {
+        fprintf(stderr, "%s: Invalid arg 'path'=%p\n", __func__, path);
+        return -1;
+    }
+
+    l3_test_log_fh = fopen(path, "w+");
+    if (!l3_test_log_fh) {
+        fprintf(stderr, "%s: Error opening log-file at '%p'. errno=%d\n",
+                __func__, path, errno);
+        perror("fopen failed");
+        return -1;
+    }
+    printf("Initialized fprintf() logging to '%s'\n", path);
+    return 0;
+}
+
+/**
+ * ****************************************************************************
+ * l3_test_log_init() - Initialize L3-logging sub-system, selecting the type of
+ * logging to be performed.
+ * ****************************************************************************
+ */
+int
+l3_test_log_init(const l3_test_log_t logtype, const char *path)
```

Comment:
I don't follow why you moved this to this file and renamed it as `l3_test_...` . That gives it a flavour that this function is either testing `log_int()` [ which it is not ]. Or, appears that this function is supposed to be under the test-cases.

But, I thought by adding fprintf(), write() and other interfaces to the Client-Server logging programs, we had sort-of agreed that maybe L3 could be extended to allow those logging interfaces, too. Yes?

If so, `l3_log_init()` was meant to the be the initialization interface.

If your objective now is to 'release' L3 only with in-memory mmap() logging, and leave fprintf() / write() purely for benchmarking purposes -- that's fine too.

Again, I am getting out-of-touch with this package and don't have too much bandwidth to "argue" and resolve this with you. So, will let you decide whichever way to go.

Reply:
Apologies for giving the impression that I'd agreed with the approach, I hadn't intended to.

I think for the library to get traction, we should make it as simple and as obvious as possible. If we want some kind of dynamic logging switching layer, I think that should be a separate GH repo. Being able to switch between them as we do for our testing/benchmarking code is super useful _for us_. But for 99% of people it will be a distraction. They just want a simple, fast logging library to use. Every bit of complexity / thing for users to have to have to learn/understand will reduce the amount the library gets used.

I've learnt all this the hard way with UDB. You have to make it soooo simple and straightforward for people to adopt. People just will not read documentation.

Reply:
I'm sure this stuff can all be made a lot better / cleaner, but the urgency is to get all our testing/benchmarking code "out of the way" so that users don't need to pollute their brainspace with it. Then I'll publish the video I have on this, and do more pushing on the socials etc (using the Undo marketing dept to do this) and hopefully we can get a little bit of traction.